From 9e0edd68d84a45b8adda281dad4238a0c3bf1f4a Mon Sep 17 00:00:00 2001
From: Roman Meshkevich <roman.meshkevich@cogentembedded.com>
Date: Wed, 28 Sep 2016 17:02:55 +0300
Subject: [PATCH] v4l: vsp1: Don't register media device when userspace API is
 disabled

The media device doesn't need to be exposed to userspace when the VSP is
fully controlled by the DU driver. Don't register it in that case.
---
 drivers/media/platform/vsp1/vsp1_drv.c | 22 +++++++++++++++-------
 1 file changed, 15 insertions(+), 7 deletions(-)

diff --git a/drivers/media/platform/vsp1/vsp1_drv.c b/drivers/media/platform/vsp1/vsp1_drv.c
index 96c2263..ea37c78 100644
--- a/drivers/media/platform/vsp1/vsp1_drv.c
+++ b/drivers/media/platform/vsp1/vsp1_drv.c
@@ -211,7 +211,8 @@ static void vsp1_destroy_entities(struct vsp1_device *vsp1)
 	}
 
 	v4l2_device_unregister(&vsp1->v4l2_dev);
-	media_device_unregister(&vsp1->media_dev);
+	if (!vsp1->info->uapi)
+		media_device_unregister(&vsp1->media_dev);
 
 	if (!vsp1->info->uapi)
 		vsp1_drm_cleanup(vsp1);
@@ -229,13 +230,20 @@ static int vsp1_create_entities(struct vsp1_device *vsp1)
 	strlcpy(mdev->model, "VSP1", sizeof(mdev->model));
 	snprintf(mdev->bus_info, sizeof(mdev->bus_info), "platform:%s",
 		 dev_name(mdev->dev));
-	ret = media_device_register(mdev);
-	if (ret < 0) {
-		dev_err(vsp1->dev, "media device registration failed (%d)\n",
-			ret);
-		return ret;
-	}
 
+	if (!vsp1->info->uapi) {
+		mdev->entity_id = 1;
+		INIT_LIST_HEAD(&mdev->entities);
+		spin_lock_init(&mdev->lock);
+		mutex_init(&mdev->graph_mutex);
+	} else {
+		ret = media_device_register(mdev);
+		if (ret < 0) {
+			dev_err(vsp1->dev, "media device registration failed (%d)\n",
+				ret);
+			return ret;
+		}
+	}
 	vsp1->media_ops.link_setup = vsp1_entity_link_setup;
 	/* Don't perform link validation when the userspace API is disabled as
 	 * the pipeline is configured internally by the driver in that case, and
-- 
2.7.4

